### ISSUE TITLE
DI: Adding a provider to a class (not component)

### ISSUE THREAD

**Issue by yfain** [angular/angular#5622]
I have a service, which is a class that internally uses Http:

```
@Injectable()
export class ProductService{

    constructor( private http: Http){}

    getProductByID(productID: string): Observable{
       return this.http.get(`/products/${productID}`)
           .map(res => res.json());
    }
}
```

To use this service, an external component must specify [HTTP_PROVIDERS]. This breaks encapsulation of my ProductService (a.k.a. Leaky Abstraction): now it's not a reusable black box but has strings attached. If I decide to replace Http with another service in the above code, the external component must modify its code as well. It would be nice if there was a way to allow a class (not a component) to specify its own providers.

Am I asking for too much?  :)

**Comment by pkozlowski-opensource** [comment]
> Am I asking for too much? :)

Yes :). If you don't want _users_ of your function to provide Http's implementation you could do: `constructor( private http: Http = new YourHttpImpl())`.

But the whole point of DI is that you say "Hey, I need _something_ that implement Http _interface_. I don't care what is the exact implementation, just give me one that makes sense for your app and fulfils the interface". As an app owner I want to have control over services used by components I'm brining in.

If you don't want people to give this control you can use my suggestion or drop DI altogether. But you can't have a cake and it a cake :-)

**Comment by PascalPrecht** [comment]
I think another thing that A2 users need to realise is that you never really instantiate a service instance as  a standalone thing. You always do it through an injector (unless you simply call `new` on something but then you also don' use DI), and injectors are only create for components, which implies you always instantiate a service by instantiating a component. And as @pkozlowski-opensource said, you service is simply asking for something of type `Http`, whereas what `Http` means really depends on what the injectors knows about this token.

So long story short, you can't have providers for services since services don't have an injector, they're just classes.

**Comment by yfain** [comment]
Well, you're explaining how Angular2 implements DI, and not how DI should work.  Forget about Http. Say we have a ComponentA with the following dependencies:
CompA <-- ServiceA <-- ServiceB <-- ServiceC 

I don't want  ComponentA to know anything about ComponentB and ComponentC.  I'll write a plunk illustrating the issue.

**Comment by pkozlowski-opensource** [comment]
@yfain I'm confused.... You can have providers for components - this is where you can define things for DI. Provide a minimal plunk as it is not clear what is your exact use-case and what coupling you see as unnecessary here.

**Comment by zoechi** [comment]
I think a service should be able to define **default** providers, which are used when no provider is specified in the parent injectors.

If a `Service` depends on `Other`, what's the point in listing `Other` in `bootstrap(AppComponent, const [Other])`?
This just bloats main with imports and and `bootstrap` with a long cumbersome to maintain list of providers.

Only if you want, for example for testing or whatever purpose, ensure `Service` uses `AlternativeOther`, **then** you add it on `bootstrap()`.

An ugly workaround could be

```dart
@Injectable()
class Service {
  Other _other;

  Service(Injector injector) {
    _other = injector.getOptional(Other);
    if (_other == null) {
      _other = new Other();
    }
    _other.doSomething();
  }
}
```
With 
```dart
  bootstrap(AppElement, [
    provide(Other, useClass: AlternativeOther)
```
`AlternativeOther` is used, without it just `Other`

**Comment by sjmf** [comment]
I think the syntax proposed by @zoechi would be practical for a use case I'm currently wrangling with. 

Consider a service (S1) which instantiates two further services (S2, S3). Using ng2's DI implementation, I provide these at the parent component level with a provider array `[S1, S2, S3]`.

It might be cleaner code to have `providers: [S1]` at my component, and allow the S1 `@Injectable` handle its own child service classes- unless I provide an alternate implementation for testing (which I do).

**Comment by MaGnaL** [comment]
Actually I was very suprised that `@Injectable` doesn't provide any possibility to add providers. Adding all providers to the `@Component` annotation (or even the bootstrap function) disagrees with my understanding of decoupled services.

Thinking of a facade service, which bundles functionality of other services, hides those complexity and provides a new and simple api. As a developer which uses this facade, I don't want to care about it's dependencies. The facade should handle them by itself.

**Comment by PascalPrecht** [comment]
> Actually I was very suprised that @Injectable doesn't provide any possibility to add providers.

Currently, `@Injectable()` is only there to force TypeScript to emit metadata. It could be any other decorator, as explained here: http://blog.thoughtram.io/angular/2015/09/17/resolve-service-dependencies-in-angular-2.html

> Adding all providers to the @Component annotation (or even the bootstrap function) disagrees with my understanding of decoupled services.

The thing is that a service doesn't get its own injector. A component does. In other words, you'd never create a service instance yourself, you'd pretty much always do it via Angular's DI which is only setup in a component.

Let's think about this for a second. We have a component tree of two components:

```
APPLICATION (Platform Injector, gets providers like HTTP_PROVIDERS, .. etc.)
COMP1 (Injector1)
     |
COMP2 (Injector2)
```

Each `Injector` can introduce its own provider configuration and "override" the one provided by the parent injector. Technically, nothing gets overridden, but rather shadowed (just like we're used to it in prototypical inheritance in JS).

Now, as soon as we inject a service in lets say COMP1, all other dependencies provided by the providers would be available to that service. Let's say we have a provider for `DataService`. This could either be done in boostrap, or on a component level. For now it's not important:

```
bootstrap(MyApp, [DataService]);
```

And let's say we have an `OnlineDataService`, which has a dependency of something of type `DataService` (note: again, `@Injectable()` is only there to emit metadata in TS):

```
@Injectable()
class OnlineDataService {
  constructor(dataService: DataService) { }
}
```

Okay cool, we can use `OnlineDataService` in our component as soon as our injector knows about that service:

```
bootstrap(MyApp, [DataService, OnlineDataService])
```

And then in the component:

```
@Component()
class COMP1 {
  constructor(ds: OnlineDataService) {

  }
}
```

Okay, so far nothing special here. But now it get's interesting:

### Let's say OnlineDataService defines its own providers

Assuming we could add providers to `@Injectable()`, it would probably look something like this:

```
@Injectable([DataService])
class OnlineDataService {
  constructor(ds: DataService) { }
}
```

Now, if `OnlineDataService` is injected into `COMP1`, where does `DataService` come from? A service doesn't have an injector, so the provider list needs to be merged with the one provided by the component. Well, then we run into the next issue, `COMP1` could've already defined a provider for `DataService`:

```
@Component({
  providers: [DataService]
})
class COMP1 { }
```

The one defined in `OnlineDataService` would now **in fact** override the one defined in `COMP1`, as `OnlineDataService` has no injector that is part of the injector tree.

A service will always share an injector with the component it is injected to and this is by design.

**Comment by zoechi** [comment]
@PascalPrecht 
`@SkipSelf()`
```ts
@Injectable([DataService])
class OnlineDataService {
  constructor(@SkipSelf() ds: DataService) { }
}
```
sounds like it fulfills that purpose. I don't know if provider resolution comes back to "self" if parents can't resolve though. 
I couldn't find more details about `@SkipSelf()`.

**Comment by PascalPrecht** [comment]
`@SkipSelf()` would skip the injector of the component where this dependency is requested and starts the lookup in the parent component, which is not what you want in that case.

**Comment by born2net** [comment]
+1 I posted the same question @ StackO as I ran into the same situation....

A service should be able to provide it's own providers...
now sure, you can do it manfully via injector.resolveAndCreateChild([...
but I would expect a more elegant solution via something like
@ProviderSelf or something of that sort...

PS

tx zoechi  for providing this discussion link,

regards

Sean

**Comment by PascalPrecht** [comment]
Again, a service doesn't have an injector. It can't have it's own providers. You surely don't want your Service to be required to have an injector property.

An Angular component knows about injectors, and that's where services are instantiated anyways.

**Comment by zoechi** [comment]
@PascalPrecht seems you don't see the demand. This prevents a lot of what DI is about in the first place. I don't say it's easy but it's quite some limitation.

**Comment by yfain** [comment]
@PascalPrecht The service is injected into some component, and it could reuse the component's injector .  The service should be able to give a provider to the component's injector. I'd call this a piggybacking feature.

**Comment by PascalPrecht** [comment]
The service *does* use the components injector. Think about what I've
written above. What if you have two services being injected on the same
level (same component), and both define providers for the same token but
with different values. Who wins?

Right. We don't know. We can't know.

**Comment by yfain** [comment]
@PascalPrecht  I recently read in a blog post by one respected Angular developer that in case of multi-providers the last one wins:

http://blog.thoughtram.io/angular2/2015/11/23/multi-providers-in-angular-2.html

Have your read this post?  :)

**Comment by PascalPrecht** [comment]
@zoechi I see the demand but I'm feeling that it's only there because it doesn't seem to be clear our A2's DI system works.

> This prevents a lot of what DI is about in the first place.

DI is about injecting dependencies into constructor, which is what we clearly do. The question is, who injects them and where are those objects created. This is where Angular's DI system comes into play.

>  I don't say it's easy but it's quite some limitation.

What exactly is the limitation?

**Comment by PascalPrecht** [comment]
> Have your read this post? :)

@yfain this is why I said the demand is only there because people don't understand the DI system or get it wrong.

> I recently read in a blog post by one respected Angular developer that in case of multi-providers the last one wins

This would 

a) require you not only to define providers on your service but also make them always multi providers. Clearly a requirement you don't want to introduce, despite we don't want to make providers a requirement of simple classes anyways, right?

b) not solve the problem as, whenever you use multiple services that have their own, but different dependencies, you'd get the wrong dependency.

This is not what multi providers are for.

**Comment by bisubus** [comment]
@PascalPrecht I'm confused a bit with a turn that the discussion has taken. I think that @yfain given a valid real-world example with `Http`.

Let's say we have self-contained module that uses some Angular injectables, no components, no views. A scenario that I've encountered in Angular 1 not once and twice, it just stands to reason:

    angular.module('someApi').service('someApi', ($http) => { ... });

In Angular 2, we should instruct the user to inject a bunch of injectables to a component manually, shouldn't we?

I understand why leaning to component-centric approach may be good for the framework, but cutting off any other option is not exactly an option that I would expect from DI that makes sense.

**Comment by hollanderbart** [comment]
+1
A service should be able to provide it's own providers. Else it would not be decoupled.

**Comment by PascalPrecht** [comment]
> angular.module('someApi').service('someApi', ($http) => { ... });

@bisubus this is exactly what you're doing in A2. In your A1 example `someApi` also doesn't define a provider for `$http`.

**Comment by ispyinternet** [comment]
+1. Seems fair, practical and logical to support this.

**Comment by bisubus** [comment]
@PascalPrecht Well, it is hard to draw an accurate analogy between A1 and A2 here, just because `$http` service was defined in `ng` module and doesn't have to be loaded explicitly - as opposed to A2 bundled injectables (HTTP_PROVIDERS).

My primary concern is unobtrusive use of bundled injectables, but I don't think that there is real difference if they are bundled or third-party injectables (e.g. OtherApi).

I would expect something like that from self-contained service in A1, so `app` module/component is clueless about `someApi` inner dependencies:

```
angular.module('someApi', ['otherApi']).service('someApi', ($http, otherApi) => { ... });
...
angular.module('app', ['someApi']).component('app', {
  controller: (someApi) => { ... })
});
```

And it looks like A2 expects it to be done this way:

```
@Injectable()
export class SomeApi {
  constructor(http: Http, otherApi: OtherApi) { }
}

export const SOME_API_PROVIDERS = [SomeApi, OtherApi, HTTP_PROVIDERS];

...

@Component({
  // luckily this
  // providers: [SomeApi, OtherApi, HTTP_PROVIDERS, STEAMING_PILE_OF_PROVIDERS]
  // can be replaced with
  providers: [SOME_API_PROVIDERS]
})
class App { }
```

So here we go again, walking in circles and maintaining a separate export as lame substitute for long-lost A1 modules.

Now, to keep this constructive and consider [the current state of things](https://github.com/angular/angular/issues/5622#issuecomment-170916680),

> A service doesn't have an injector, so the provider list needs to be merged with the one provided by the component.

That's correct. I would consider

```
@Injectable([HTTP_PROVIDERS])
export class SomeApi { }

...

@Component({
  providers: [SomeApi]
})
class App { }
```

a nice syntactic sugar for

```
@Injectable()
export class SomeApi { }
export const SOME_API_PROVIDERS = [SomeApi, HTTP_PROVIDERS];

...

@Component({
  providers: [SOME_API_PROVIDERS]
})
class App { }
```

> A service will always share an injector with the component it is injected to and this is by design.

Letting a service to just get its hands on component's injector would be good enough, IMO.
Having something like
```
constructor(@Inject(Vehicle) vehicle, @InjectChild(Car, Engine) [car, engine]) { }
```
instead of dealing with `Injector` directly would be interesting, but it is slightly out of the current scope, I guess.

**Comment by KlausConrado** [comment]
+1
Same situation here, i created a service interface and a http implementation that depends on http providers. Injecting the http implementation in my component should be enough. It's strange the component be forced to inject a dependency of his dependency. As said above, this breaks the coupling.

**Comment by schippie** [comment]
I have resorted for this to bundle all providers i require for a service in a constant that i export. That constant is then the thing i import into my components / bootstrap method. Its a little dirty since some have duplications and i have no clue what Angular does with redeclaration of the same provider.

**Comment by zoechi** [comment]
@schippie When adding a provider for the same key (type) multiple times, Angular just takes the one added last. This makes it easy to for example get mocks injected:
```
[HTTP_PROVIDERS, MockBackend, provide(XHRBackend, {useExisting: MockBackend})]
```
to make DI inject `MockBackend` where `XHRBackend` is requested (which is part of `HTTP_PROVIDERS`

**Comment by weddell** [comment]
I just came across this and couldn't believe I needed to add a dependents dependencies, assuming I must have been doing it wrong.

Doesn't this go against fundamentals like decoupling, abstraction, encapsulation. 

When you use, say gulp, you don't have to import it's dependencies like vinyl. What is the point of facades if you need to know - and look at what it's dependencies are. Let alone if you had a chain of services built on top of each other. 

Agree also you are discussing "how" angular works, that's all well and good, but it's a flaw IMO. Just because you design something someway doesn't mean it wasn't a design flaw.

**Comment by schippie** [comment]
@weddell Not if you go about it like @zoechi specified. If you specify a list of providers your service depends on you can easily decouple it like so:

/some-services/index.ts
```javascript
import { Http }  from "@angular/http";
import { SomeService } from "./some-service.ts";

export const SERVICE_PROVIDERS= [
   Http,
   SomeService
]
```

/some-services/some-service.ts
```javascript
import { Injectable }  from "@angular/core";
import { Http }  from "@angular/http";

@Injectable()
export class SomeService {
    constructor(private http: Http) { .. }
}
```

Also see the part about barrels: https://angular.io/styleguide#!#create-and-import-barrels

**Comment by kjozic** [comment]
I am new Angular2 user so excuse me if I got something wrong.

If I get it right, there is a global injector (via bootstrap() ) and a component local injector (via Component({ providers: [...]}) ). The global injector provides singleton services (one instance of a service per application). Each component has its own injector which provides services local to that component only. In the case of injection, component injector wins over global injector. Is this right?

So we have this situation:
```
@Injectable()
class ServiceA {}

bootstrap(..., [ServiceA]);

@Component ({
providers: [ServiceA]
})
class ComponentA {
    constructor(serviceA:ServiceA) {
        //My injector wins over global
    }
}
```
If this is right, I don't understand why it is not possible to continue this practice with Injectable({providers: []})?
If we declare provider inside of a service it should take precedence over component and global.

Let's look at following C example:
```
int main(void) {
    int a = 1;    //lets say this is bootstrap()
    printf("%i\n", a);
    {
        int a = 2;    //this is ComponentA
        printf("%i\n", a);
        {
            int a = 3;    //and this is ServiceB
            printf("%i\n", a);
        }
    }
}
```
Printout is:
```
1
2
3
```
This example shows that innermost variable takes precedence.

So if we have ServiceA in bootstrap(), ComponentA and ServiceB, ServiceB (as innermost) should use ServiceA provided by its own injector.

**Comment by KostyaTretyak** [comment]
@yfain, how about this:

```ts
import 'reflect-metadata';
import {ReflectiveInjector} from '@angular/core/src/di';

@Injectable()
export class ProductService
{
    private http;

    constructor()
    {
      let injector = ReflectiveInjector.resolveAndCreate( [HTTP_PROVIDERS] );
      this.http = injector.get(Http);
    }

    getProductByID(productID: string): Observable
    {
       return this.http.get(`/products/${productID}`)
           .map(res => res.json());
    }
}
```

Now you can use this without constructor's params:

```ts
let productService = new ProductService;
```

*Notice:  without `import 'reflect-metadata';` we have error*

**Comment by yfain** [comment]
The whole idea of DI is to avoid using the new operator. I could as well
create an instance of Http using new inside the ProductService.

I'd like to see DI working with regular decorated classes.

**Comment by zoechi** [comment]
@yfain https://github.com/angular/angular/issues/9726 might go into the right direction.

**Comment by PascalPrecht** [comment]
@KostyaTretyak creating an injector manually causes it *not* to be part of the injector tree. That's not an option. Angular calls this code for you.

@zoechi I don't think this is related.

**Comment by KostyaTretyak** [comment]
@PascalPrecht, but how to use DI as `standalone module`:

>Angular ships with its own dependency injection framework. This framework can also be used as a standalone module by other applications and frameworks.

I really like the idea @Injectable, and try to use it in node.js. I was browsing [di.js](https://github.com/angular/di.js), but it seems that it is no longer maintain more.

I would be grateful to you if you show how I can use DI in node.js with decorator @Indectable

**Comment by akash-wadhwa** [comment]
I think this is basic requirement for DI where services should be injected into another services without providing providers at the component level. This basically break the decoupling.
totally agree with @yfain.
@PascalPrecht is explaining how angular DI works not how DI works.

**Comment by bisubus** [comment]
Fortunately, `angular.module` came back in the guise of `NgModule`, no more BARREL_ROLLS.

**Comment by zoechi** [comment]
NgModule is not related to this issue

**Comment by born2net** [comment]
I was going to say the same thing :)

**Comment by vicb** [comment]
Are you suggesting that I wrongly closed the issue ?

The initial report is not an Angular issue according to me

**Comment by born2net** [comment]
@zoechi on unrelated matter, can you give me your opinion on:
https://github.com/angular/angular/issues/11904#issuecomment-249432081

**Comment by zoechi** [comment]
@vicb if @bisubus comment was the reason you closed the issue, it was at least closed for the wrong reason ;-)

**Comment by bisubus** [comment]
@zoechi I hope it didn't! However, it appears to me that NgModule isn't the same thing as the title says , but it perfectly covers both the case from the original post and [this one](https://github.com/angular/angular/issues/5622#issuecomment-193302391). If there were valid reasons for the subject, now there's one less.

**Comment by zoechi** [comment]
NgModule just adds the providers to the root injector. That's not what the issue is about. The issue is about some injector hierarchy within services. Lazy loaded modules come close though.

**Comment by vicb** [comment]
@bisubus comment triggered a notification and I took a look at the issue.

**Comment by Squiggle** [comment]
I have a situation where a service - **AuthenticatedHttpClient** - needs to use a different implementations of na **AuthTokenResolver** depending on the configuration.

Currently the only two solutions (as far as my understanding goes) are:
1. On each component that consumes the AuthenticatedHttpClient, implement the provider that decides which AuthTokenResolver instance to use.
2. In the constructor of the AuthenticatedHttpClient, consult the configuration and explicitly create the correct AuthTokenResolver, i.e. `this.tokenResolver = config.UseTokenResolverA ? new TokenResolverA() : new TokenResolverB();`

This was simple in Angular 1, where I could just define a custom provider which would be used throughout all services (not just directives and components).

I have read through this thread and it seems some people are talking cross-purposes.
Can someone clarify whether or not this behavior is intentionally missing? And if so, what the preferred approach now is?

**Comment by wpitallo** [comment]
Hi Everyone,

I was having a similar issue. I managed to finally get it working using the following approach since some of the solutions here use HTTP_PROVIDERS that has been depreciated. What I would like to know is if there is any downside or concerns about the way this has been implemented (this.http in my service particularity). I did try injecting http into the service constructor but because I am not using a component I kept getting: No Provider for http

Any feedback would be greatly appreciated.

Service:


import {Observable} from 'rxjs';
import 'rxjs/add/operator/map'
import {helperService} from '{{rootPath}}CodeScience.Client.Global/helpers/helper.services';
import {LOCATION_MODEL} from '../models/LOCATION_MODEL';
import { Http, Response,RequestOptions, URLSearchParams } from '@angular/http';
import { ReflectiveInjector,Injectable } from '@angular/core';
import { Http, XHRBackend, ConnectionBackend, BrowserXhr, ResponseOptions, XSRFStrategy, BaseResponseOptions, CookieXSRFStrategy, RequestOptions, BaseRequestOptions } from '@angular/http';


@Injectable()
export class LocationService {

  constructor() {
    this.http =  ReflectiveInjector.resolveAndCreate([
      Http, BrowserXhr,
      { provide: ConnectionBackend, useClass: XHRBackend },
      { provide: ResponseOptions, useClass: BaseResponseOptions },
      { provide: XSRFStrategy,  useValue: new CookieXSRFStrategy('csrftoken', 'X-CSRFToken')},
      { provide: RequestOptions, useClass: BaseRequestOptions }
    ]).get(Http);
  }

LoadObjectByKey(inLOCATION_DETAILS_KEY: string, inLoadLinkedObjects: boolean) : Observable<LOCATION_MODEL> {

                        let params: URLSearchParams = new URLSearchParams();
                        params.set('LOCATION_DETAILS_KEY', inLOCATION_DETAILS_KEY);
                        params.set('inLoadLinkedObjects', inLoadLinkedObjects);
                        let options = new RequestOptions({headers: helperService.getHeaders(), search: params});

                        return this.http.get(helperService.getBaseUrl() + 'api/bootstrap/LOCATION/LoadObjectByKey', options)
                            .map((response: Response) => response.json(), error => this.error = error.json())
                            .catch((error:any) => helperService.handleError(error,Response));
}


}

Class file:

import {ValidationService} from '{{rootPath}}CodeScience.Client.Global/services/validation.service';
import { FormGroup, FormBuilder, Validators } from '@angular/forms';
import { ReflectiveInjector } from '@angular/core';
import * as SystemApiService from '{{rootPath}}core/services/_api.services';

export class LOCATION_MODEL { 

  _LOCATION_MODEL_FORM: FormGroup;

  constructor(){

    let injector = ReflectiveInjector.resolveAndCreate([FormBuilder,SystemApiService.LocationService]);
    this.formBuilder = injector.get(FormBuilder);

    this.locationService =  injector.get(SystemApiService.LocationService);

  LoadObjectByKey(key){
    this.locationService.LoadObjectByKey(key).subscribe((result) => this.LOCATION_MODEL = result,       console.log(this.LOCATION_DETAILS_MODEL_OBJECT));
  }


}

**Comment by zoechi** [comment]
@wpitallo 
GitHub issues are for bug reports and feature requests.
    For support questions please use other channels like the ones listed in [CONTRIBUTING - Got a Question or Problem?](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question)
